### ISSUE TITLE
[mysql] Collect size of databases from FS point of view

### ISSUE THREAD

**Issue by iranzo** [sosreport/sos#1144]
It would be great to get the output of:

du -s /var/lib/mysql/* included in sosreports as it should give a good idea if some database is getting out of control (for example keystone tokens or another)

**Comment by jcastill** [comment]
Hello Pablo, 

I'm not really sure about the 'du' approach. It may take a long time to run and cause issues with sosreport taking too long if the directory is huge. An alternative could be to issue a 'mysql' command to show the size of the databases, but I'm not convinced this is the best way to go. 
So a second alternative could be to run 'ls' with options like '-laR' or '-laRZ' and capture the output, adding even '-h' if we want somethign more human readable. Could this work? Or do you prefer 'du' for a specific reason? 
If you are ok with this approach, I have a patch almost ready to send.

**Comment by iranzo** [comment]
@jcastill Was this finally submitted?

Thanks!
Pablo

**Comment by jcastill** [comment]
@iranzo I started doing some tests with the solutions I proposed but I realized it does not do what you actually want directly in the way that 'du' would do. What comes to mind is to get the result from ls and sum it to offer a total size of the databases. But then, customer could always do it by themselves easily as well, instead of us adding the maths to the plugin. 
Ping @pmoravec what do you think fits more with sos' philosophy? Let me know if anything is unclear.

**Comment by iranzo** [comment]
The problem here is not the customer doing the math, but us being able to detect oversized databases that indicate errors for example in keystone token purge or ceilometer expirations for instance.

**Comment by bmr-cymru** [comment]
You can do it with `ls -s` but then it's hard to see how that is superior to using `du`.

Are there no database commands that can answer this question directly?

**Comment by jcastill** [comment]
From what I've seen, we don't have any mysql command we can use in the command line. We could run a mysql query which is basically  a SELECT followed by a sum, i.e. something like:

SELECT table_schema "database", sum(data_length + index_length)/1024/1024/1024 "size in GB" FROM information_schema.TABLES GROUP BY table_schema;

But somehow I assumed that we preferred not to do a mysql query, sorry if my assumption is wrong.
So that leaves us with:
- A mysql query. Not sure if we should go this way, because we'll need to provide user and pass details as far as I know.
- 'du', but this may take too long in big directories and cause issues.
- 'ls -s', I think may be quicker. 

@iranzo another question: What will be an 'oversized' database? 
And another though: Lets imagine that customer has more than one DB, completely unrelated to keystone or ceilometer. Wouldn't getting 'du' or similar show an incorrect view of the issue by telling you the total space used, compared to than 'ls', which could tell you the size of each and every database, and then you can decide is certain db is oversized or not?

**Comment by iranzo** [comment]
Mysql could be ok, but as you said, requires username and password so it's usually easier to do 'du'

If the customer has another database in the environment that it's not OSP related and it's an OSP case, that's an ENOTSUPPORTED case ;-)

if we do use `du -s /var/lib/mysql/*` we do see size for each database separatedly:

~~~
[root@controller-0 mysql]# time du -s *
412     aodh
16      aria_log.00000001
4       aria_log_control
4468    cinder
131076  galera.cache
2256    glance
5248    gnocchi
4       grastate.dat
4       gvwstate.dat
2428    heat
18432   ibdata1
5120    ib_logfile0
5120    ib_logfile1
4776    keystone
1016    mysql
0       mysql.sock
14304   nova
4532    nova_api
14192   nova_cell0
4       nova_placement
24476   ovs_neutron
812     panko
212     performance_schema

real    0m0.014s
user    0m0.001s
sys     0m0.011s
~~~

In this case, ignore the timing, and sizes as this is a recently deployed environment.

If we can get the ls output with full path, so we can do a 'postprocessing' of that output and do the math on our side, that would be ok as well.

**Comment by jcastill** [comment]
@iranzo at the end I used your original suggestion. Here is the result I got in a test sosreport:

[root@sos-mysql-test sosreport-sos-mysql-du-pr-20180208090916]# ll sos_commands/mysql/
total 4
-rw-r--r--. 1 root root 411 Feb  8 09:09 du_-s_.var.lib.mysql
[root@sos-mysql-test sosreport-sos-mysql-du-pr-20180208090916]# cat sos_commands/mysql/du_-s_.var.lib.mysql 
1104	/var/lib/mysql/mysql
4542516	/var/lib/mysql/test
4	/var/lib/mysql/aria_log_control
16	/var/lib/mysql/aria_log.00000001
77824	/var/lib/mysql/ibdata1
49152	/var/lib/mysql/ib_logfile1
49152	/var/lib/mysql/ib_logfile0
4	/var/lib/mysql/performance_schema
4	/var/lib/mysql/mysql_upgrade_info
24	/var/lib/mysql/tc.log
0	/var/lib/mysql/mysql.sock
0	/var/lib/mysql/multi-master.info
182536	/var/lib/mysql/employees

Does it look like you need?

**Comment by iranzo** [comment]
Yes, in that way we can check which ones are affected, in the case of keystone for example it will be the ibdata1 and the keystone database for example

Thanks jcastillo++

**Comment by iranzo** [comment]
@jcastill could be expect the resulting file in the sosreport to be <sos_commands/mysql/du_-s_.var.lib.mysql> ? if so I can start coding some checks on our side in case sosreport is updated and grabbing the data

**Comment by jcastill** [comment]
@iranzo yes, that's the full path.

**Comment by iranzo** [comment]
Thanks a lot, as this enables us to start working on plugins for citellus (https://github.com/zerodayz/citellus)  that take that into account.

Great!
Pablo
